### ISSUE TITLE
Is there any plan to integrate NCCL 2.0 or horovod into CollectiveAllReduceStrategy?

### ISSUE THREAD

**Issue by wangsiyu** [tensorflow/tensorflow#24505]
**System information**
- TensorFlow version (you are using): master
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
`Horovod` brings significant performance to distribution execution especially when using `NCCL` to do `all_reduce`. Is there any plan to do integration  or simply support NCCL 2.0 in `CollectiveAllReduceStrategy`? I am willing to contribute for it!

**Will this change the current api? How?**
No，it will be another `cross_device_ops` for candidate.

**Who will benefit with this feature?**
Anyone who wants to use `CollectiveAllReduceStrategy` to improve performance.

**Any Other info.**
Currently I am working for distribution optimization and I hope to have a discussion. Thanks.

@yuefengz

**Comment by yangjunpro** [comment]
Thank Siyu for submitting this issue.

@yuefengz  We are willing to contribute to Distribution Strategy to make it greater and brings more benefits to end-users. Currently, we found Horovod is a nice backend to be used to support CollectiveAllreduceStrategy and we are willing to work on it. Any suggestions or feedback from your side are highly welcome.

Thanks

**Comment by byronyi** [comment]
I think @dubey is working on this, and you could take a look at 681f6a6ac9344425549c8a38fcbfae854e7deddf. 

It seems to be rolled back in current master though. I expect it to come back later on.

**Comment by yuefengz** [comment]
@yangjunpro Sorry I missed this thread from my inbox somehow. @dubey is working on the nccl implementation, we hope to have it soon. It bypasses grpc and should give `CollectiveAllReduceStrategy` speed-up. 

If you are interested in contributing, it is very welcomed. We can discuss what you want to add to Distribution Strategy and we can work on a RFC together.

**Comment by wangsiyu** [comment]
@yuefengz @byronyi Thanks for reply ! I will take a look at this commit soon and then have a discussion with you.

**Comment by byronyi** [comment]
Tensor fusion is implemented with `ScopedAllocator` in 380a032e64222db73766ef18ea51130e3df4c15f, and hierarchical all-reduce is implemented with hierarchical `Broadcaster` in d4bb99250a2f8d15985823baa5ccfbc55998f7bb. Those are with native TF collective ops and may not related to NCCL though. @dubey might have some ideas.

**Comment by yangjunpro** [comment]
> Tensor fusion is implemented with `ScopedAllocator` in [380a032](https://github.com/tensorflow/tensorflow/commit/380a032e64222db73766ef18ea51130e3df4c15f), and hierarchical all-reduce is implemented with hierarchical `Broadcaster` in [d4bb992](https://github.com/tensorflow/tensorflow/commit/d4bb99250a2f8d15985823baa5ccfbc55998f7bb). Those are with native TF collective ops and may not related to NCCL though. @dubey might have some ideas.

@byronyi 

Thanks for the sharing. I have gone through the ScopedAllocator implementation and it looks that it currently plays as a bridge between a large tensor and its corresponding smaller slices.  To a certain degree it can be viewed as "Tensor Fusion". However from distributed optimization perspective, it may not be complete, since during the execution of distribution communication flow, it may not be regarded as a fixed "fusion pattern" in which tensors should be fused together. Background workloads, currently data pattern may both impact the Tensor fusion behavior.  
Also is there any place in TF community's branch code already using the ScopedAllocator? I would like to take a look at its concrete use cases.

Thanks.

**Comment by byronyi** [comment]
You could take a look at 0b522fd22b986704d1056254961cc7988ae182eb, enabled by 282750fee5e2df502436ca9ef6a95283f8adab34 and 09e529ff5adb916e40481563698dee72e8a15162. It is enabled by [this toggle](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/core/protobuf/rewriter_config.proto#L77).

**Comment by yangjunpro** [comment]
> You could take a look at [0b522fd](https://github.com/tensorflow/tensorflow/commit/0b522fd22b986704d1056254961cc7988ae182eb), enabled by [282750f](https://github.com/tensorflow/tensorflow/commit/282750fee5e2df502436ca9ef6a95283f8adab34) and [09e529f](https://github.com/tensorflow/tensorflow/commit/09e529ff5adb916e40481563698dee72e8a15162). It is enabled by [this toggle](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/core/protobuf/rewriter_config.proto#L77).

Thanks for sharing the details. I will take a look at them.

**Comment by yangjunpro** [comment]
> > Thanks for the quick responses, @byronyi! As Yuefeng said, we welcome contributions and I'd love to hear more about your plans. Please find my thoughts below on the discussion until now.
> > 
> > * After [`f6b81f4`](https://github.com/tensorflow/tensorflow/commit/f6b81f458da1f7b68e06c6a4501c5c380d33d590), the key pieces for enabling NCCL with CollectiveAllReduceStrategy are now in TensorFlow.  We're running some final tests, and we will soon provide a way for you to test this out.  It would be great if you can try it out.
> > * The goal of `ScopedAllocator` is similar to Tensor Fusion: convert many small all reduces into fewer, larger all reduces.  `ScopedAllocator` is already a part of `CollectiveOps` implementation, so those benefits should carry over to the NCCL implementation as well.
> > * We do indeed have a hierarchical broadcast implementation, and it would be possible to build something that combines the ideas in [`d4bb992`](https://github.com/tensorflow/tensorflow/commit/d4bb99250a2f8d15985823baa5ccfbc55998f7bb) and [`f6b81f4`](https://github.com/tensorflow/tensorflow/commit/f6b81f458da1f7b68e06c6a4501c5c380d33d590) to get a hierarchical NCCL-based all reduce implementation.  I've thought about this in the past but haven't gotten around to doing it yet.
> > 
> > > However from distributed optimization perspective, it may not be complete, since during the execution of distribution communication flow, it may not be regarded as a fixed "fusion pattern" in which tensors should be fused together. Background workloads, currently data pattern may both impact the Tensor fusion behavior.
> > 
> > 
> > I'd like to understand more what you mean by this. Can you elaborate, or perhaps provide an example?
> 
> I think his mean that the tensor fusion buffer size cannot be auto tuned online in every training loop. Is that correct ? @yangjunpro

Yep, this is what I mean. For different training iteration, due to the background workload or other production environment variation, the best tensor fusion buffer size may also change dynamically.

**Comment by byronyi** [comment]
@yangjunpro That would be very interesting. Do you have any further reference/experiment to illustrate the potential gain? AFAIK, one could use `contrib/all_reduce` for prototyping when tuning the best merge/split strategy, but you may have different choices.

See https://github.com/tensorflow/benchmarks/blob/master/scripts/tf_cnn_benchmarks/allreduce.py#L448 as an example.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by yuefengz** [comment]
You can now use `tf.distribute.MultiWorkerMirroredStrategy(communication=tf.distribute.experimental.CollectiveCommunication.NCCL`)` to try out our nccl implementation. Nccl is faster on some networks than our previous implementation but allocates extra memory buffers.

**Comment by poxvoculi** [comment]
Horovod's TensorFusion is a dynamic technique: During execution nodes report on the tensor reductions that are ready to go, then a master determines the next reduction set based on an intersection of those reported ready.  I assume there is next a compacting data copy at each node before the collective op begins (which is an opportunity for compaction or precision reduction).  All of this requires some out of band communication to a master which potentially introduces extra latency or high in-cast at large scale, plus there's also extra latency for data copies.  Depending on the computation it may be possible to overlap much of this latency with other useful work.

ScopedAllocator is a static technique: upstream operations compute into adjacent regions of a pre-allocated tensor.  The collective op can begin as soon as the final inbound edge completes, there is no out-of-band communication necessary.  As you point out, it's not clear that static analysis can find the best strategy, or that the best strategy is consistent from step to step.

It seems an empirical question which technique will work best for a particular graph in a particular distribution.  The TF team is interested in ways of improving the application of ScopedAllocator, perhaps by periodically examining runtime performance to modify the merge strategy.

**Comment by wangsiyu** [comment]
@poxvoculi So is there any plan to do the pre-analysis based on cost model in static tensor fusion technique or other tuning method?

**Comment by poxvoculi** [comment]
@wangsiyu I've had some casual conversations with colleagues about the feasibility of doing something like that, but there's no concrete plan to do anything yet.  I think it would be interesting to do some detailed performance comparisons of the dynamic Horovod approach with the static ScopedAllocator approach and see where each looks stronger.  That might give some insight into the most worthwhile next project in this area.

**Comment by yangjunpro** [comment]
@poxvoculi 

Thanks for the feedback.

@yuefengz 

Internally we have already integrated Horovod(we have gotten rid of MPI to ensure as less extra dependencies to be introduced as possible ) into CollectiveAllReduceStrategy and benchmarked with some of our in-house models. The performance also looks promising. For a Transformer NLP model, the rough number is as following: for 2node 16 GPUs, around 14X speed-up against single GPU version; for 3node 24 GPUs, around 18X speed-up, with only 10Gbps ethernet device. Further performance improvement is expected if better hardware is utilized, such as RDMA or IB, etc. 

We are considering whether to make this Horovod integration a separate PR into distribution strategy. Any feedback from your folks are highly welcome. 

Thanks.

**Comment by byronyi** [comment]
@yangjunpro We are having this discussion on collective communication and networking API in general within the networking SIG. If you are interested, feel free to send us a design draft to networking@tensorflow.org. We are also having a conference call this Wednesday morning (7am GMT+8), and it is a great place to escalate the discussion.

**Comment by yangjunpro** [comment]
@byronyi 

Sure, we will join the conference this time.

Is there any dial-in number or Hangout invitations?

Thanks

**Comment by byronyi** [comment]
You should be able to see the invitation once you add yourself to https://groups.google.com/a/tensorflow.org/forum/m/?pli=1#!forum/networking

**Comment by dubey** [comment]
Closing issue; NCCL 2.0 is now integrated into MultiWorkerMirroredStrategy.
